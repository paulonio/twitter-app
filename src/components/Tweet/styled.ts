import { styled } from 'styled-components';

export const TweetWrapper = styled.div`
  width: 100%;
  max-width: 910px;
  padding: 10px 35px 21px;
  display: flex;
  border-top: solid 1px ${({ theme }) => theme.colors.border};
`;

export const TweetAvatar = styled.div`
  width: 49px;
  height: 54px;
`;

export const TweetContent = styled.div`
  width: 93%;
  max-width: 747px;
  margin-left: 8px;
`;

export const TweetText = styled.p`
  margin-top: 5px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
`;

export const ImageWrapper = styled.div`
  width: 100%;
  max-width: 679px;
  height: auto;
  max-height: 453px;
  margin-top: 15px;
  border-radius: 20px;
  overflow: hidden;
`;

export const TweetImage = styled.img`
  width: 100%;
  height: 100%;
`;

export const TweetLike = styled.div`
  width: 6%;
  max-width: 53px;
  margin-top: 22px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  &:hover {
    cursor: pointer;
  }
`;

export const LikeIcon = styled.img.attrs({ src: '/assets/icons/Like.svg', alt: 'Like icon' })`
  width: 24px;
  height: 24px;
`;

export const LikeAmount = styled.span`
  margin-left: 10px;
  font-size: 16px;
  line-height: 21px;
  font-weight: 400;
  color: ${({ theme }) => theme.colors.text};
  opacity: 0.6;
`;
