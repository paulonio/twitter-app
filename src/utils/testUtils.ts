import { combineReducers, configureStore, PreloadedState } from '@reduxjs/toolkit';
import authReducer from '@store/slices/authSlice';
import tweetReducer from '@store/slices/tweetSlice';
import themeReducer from '@store/slices/themeSlice';

const rootReducer = combineReducers({
  auth: authReducer,
  tweet: tweetReducer,
  theme: themeReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
