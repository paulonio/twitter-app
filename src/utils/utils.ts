import {
  doc,
  DocumentSnapshot,
  DocumentData,
  getDoc,
  collection,
  updateDoc,
  arrayUnion,
} from 'firebase/firestore';
import { db } from '../../firebase';
import { TweetType } from '../store/slices/tweetSlice';
import { User } from '../store/slices/authSlice';

export const getDocument = async (column: string, document: string) => {
  const docRef = doc(db, column, document);
  const docSnap: DocumentSnapshot<DocumentData> = await getDoc(docRef);
  return docSnap;
};

export const updateDocument = async (column: string, document: string, payload: TweetType) => {
  const twitterRef = collection(db, column);
  await updateDoc(doc(twitterRef, document), { tweets: arrayUnion(payload) });
  const docSnap: DocumentSnapshot<DocumentData> = await getDocument(column, document);
  return docSnap;
};

export const updateUsers = async (column: string, document: string, payload: User | User[]) => {
  const twitterRef = collection(db, column);
  if (payload instanceof Array) {
    await updateDoc(doc(twitterRef, document), { users: payload });
  } else {
    await updateDoc(doc(twitterRef, document), { users: arrayUnion(payload) });
  }
  const docSnap: DocumentSnapshot<DocumentData> = await getDocument(column, document);
  return docSnap;
};

export const isUserArray = (value: unknown): value is User[] => {
  if (!Array.isArray(value)) {
    return false;
  }

  return value.some((item) => !!item.uid);
};

export const isTweetArray = (value: unknown): value is TweetType[] => {
  if (!Array.isArray(value)) {
    return false;
  }

  return value.some((item) => !!item.tweet);
};
