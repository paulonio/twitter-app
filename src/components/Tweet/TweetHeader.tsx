import React, { FC } from 'react';

import type { TweetType } from '@store/slices/tweetSlice';

import { Nickname, Username } from '@components/Menu/styled';
import { TweetHeaderWrapper, TweetTitle, EditButton, Edit } from './styled';

type TweetHeaderProps = Pick<TweetType, 'displayName' | 'userEmail' | 'date'>;

const TweetHeader: FC<TweetHeaderProps> = ({ displayName, userEmail, date }) => {
  return (
    <TweetHeaderWrapper>
      <TweetTitle>
        <Username>{displayName}</Username>
        <Nickname>
          @{userEmail.split('@')[0]} Â· {date}
        </Nickname>
      </TweetTitle>
      <EditButton>
        <Edit />
      </EditButton>
    </TweetHeaderWrapper>
  );
};

export default TweetHeader;
