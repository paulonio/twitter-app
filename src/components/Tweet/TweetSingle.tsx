import React from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import Tweet from '@components/Tweet/Tweet';

import type { StoreType } from '@store/index.ts';
import type { TweetType } from '@store/slices/tweetSlice';

import { TweetSingleWrapper } from './styled';

const TweetSingle = () => {
  const tweets = useSelector<StoreType, TweetType[]>((state) => state.tweet.tweets);
  const { id } = useParams();
  const [currentTweet] = tweets.filter((tweet) => tweet.id === id);

  return (
    <TweetSingleWrapper>
      <Tweet currentTweet={currentTweet} />
    </TweetSingleWrapper>
  );
};

export default TweetSingle;
