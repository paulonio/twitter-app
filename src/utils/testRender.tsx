import React, { FC } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import Theme from '@src/components/Theme/Theme';
import { persistor } from '@src/store';
import { BrowserRouter } from 'react-router-dom';
import { PersistGate } from 'redux-persist/integration/react';
import { setupStore, AppStore, RootState } from './testUtils';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

interface WrapperProps {
  children: JSX.Element | JSX.Element[] | string | string[];
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper: FC<WrapperProps> = ({ children }): JSX.Element => {
    return (
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <BrowserRouter>
            <Theme>{children}</Theme>
          </BrowserRouter>
        </PersistGate>
      </Provider>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
