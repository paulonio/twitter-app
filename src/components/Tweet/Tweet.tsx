import React, { FC } from 'react';

import Avatar from '@components/Avatar/Avatar';
import TweetHeader from '@src/components/Tweet/TweetHeader';
import ErrorBoundary from '@components/ErrorBoundary/ErrorBoundary';

import type { TweetType } from '@store/slices/tweetSlice';

import LikeIcon from '@icons/LikeIcon';
import {
  ImageWrapper,
  LikeAmount,
  TweetAvatar,
  TweetContent,
  TweetImage,
  TweetLike,
  TweetText,
  TweetWrapper,
} from './styled';

interface TweetProps {
  currentTweet: TweetType;
}

const Tweet: FC<TweetProps> = ({ currentTweet }) => {
  const { tweet, userEmail, displayName, urlToImage, date } = currentTweet;

  return (
    <ErrorBoundary>
      <TweetWrapper data-testid="tweet-item">
        <TweetAvatar>
          <Avatar />
        </TweetAvatar>
        <TweetContent>
          <TweetHeader userEmail={userEmail} displayName={displayName} date={date} />
          <TweetText data-testid="tweet-text">{tweet}</TweetText>
          {urlToImage && (
            <ImageWrapper>
              <TweetImage src={urlToImage} alt="Tweet" />
            </ImageWrapper>
          )}
          <TweetLike>
            <LikeIcon />
            <LikeAmount>8</LikeAmount>
          </TweetLike>
        </TweetContent>
      </TweetWrapper>
    </ErrorBoundary>
  );
};

export default Tweet;
