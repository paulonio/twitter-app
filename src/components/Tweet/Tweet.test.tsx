import React from 'react';
import { screen } from '@testing-library/react';
import { renderWithProviders } from '@src/utils/testRender';
import type { TweetType } from '@src/store/slices/tweetSlice';
import Tweet from './Tweet';

const tweetString =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.';

export const tweetData: TweetType = {
  tweet: tweetString,
  userEmail: 'test@test.com',
  displayName: 'test',
  date: 'Jun 4',
  userUid: '111111111',
  id: '222222222222',
};

describe('Tweet', () => {
  it('should render tweet component', () => {
    const { container } = renderWithProviders(<Tweet currentTweet={tweetData} />);

    expect(container).toMatchSnapshot();
  });

  it('should display data correctly', () => {
    renderWithProviders(<Tweet currentTweet={tweetData} />);

    const tweet = screen.getByTestId('tweet-text');
    const username = screen.getByTestId('username');
    const nickname = screen.getByTestId('nickname');

    expect(tweet.textContent).toEqual(tweetString);
    expect(username.textContent).toEqual('test');
    expect(nickname.textContent).toEqual('@test Â· Jun 4');
  });
});
