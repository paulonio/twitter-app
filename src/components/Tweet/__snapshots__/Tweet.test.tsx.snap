// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet should render tweet component 1`] = `
<div>
  <div
    class="sc-jySWPz dhyrvK"
    data-testid="tweet-item"
  >
    <div
      class="sc-burPHO eXNuIf"
    >
      <img
        alt="Avatar"
        class="sc-cHausL hIINSS"
        src="/assets/images/Avatar.svg"
      />
    </div>
    <div
      class="sc-hlwzUd TjVjh"
    >
      <div
        class="sc-XDmMx fZWpeS"
      >
        <div
          class="sc-jjgsYe coYsJH"
        >
          <h4
            class="sc-cEPTvB knBcr"
            data-testid="username"
          >
            test
          </h4>
          <div
            class="sc-fPKvMM jappeS"
            data-testid="nickname"
          >
            @
            test
             Â· 
            Jun 4
          </div>
        </div>
        <button
          class="sc-hoAVFr bxqzGG"
        >
          <svg
            class="sc-hyPqDM dtxtnG"
            fill="currentColor"
            focusable="false"
            height="4"
            role="img"
            viewBox="0 0 18 4"
            width="18"
          >
            <path
              d="M4.67308 2C4.67308 3.10457 3.79487 4 2.71154 4C1.62821 4 0.75 3.10457 0.75 2C0.75 0.895431 1.62821 0 2.71154 0C3.79487 0 4.67308 0.895431 4.67308 2Z"
            />
            <path
              d="M11.2115 2C11.2115 3.10457 10.3333 4 9.25 4C8.16667 4 7.28846 3.10457 7.28846 2C7.28846 0.895431 8.16667 0 9.25 0C10.3333 0 11.2115 0.895431 11.2115 2Z"
            />
            <path
              d="M17.75 2C17.75 3.10457 16.8718 4 15.7885 4C14.7051 4 13.8269 3.10457 13.8269 2C13.8269 0.895431 14.7051 0 15.7885 0C16.8718 0 17.75 0.895431 17.75 2Z"
            />
          </svg>
        </button>
      </div>
      <p
        class="sc-kwrclo jrfDXY"
        data-testid="tweet-text"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </p>
      <div
        class="sc-goxvPd ftfERC"
      >
        <svg
          fill="currentColor"
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 25 24"
          width="25"
        >
          <path
            d="M12.8 21.914H12.786C10.203 21.866 2.75 15.132 2.75 8.754C2.75 5.69 5.275 3 8.153 3C10.443 3 11.983 4.58 12.799 5.73C13.613 4.582 15.153 3 17.444 3C20.324 3 22.848 5.69 22.848 8.755C22.848 15.131 15.394 21.865 12.811 21.912H12.8V21.914ZM8.154 4.501C6.074 4.501 4.251 6.489 4.251 8.756C4.251 14.496 11.285 20.352 12.801 20.414C14.319 20.352 21.351 14.497 21.351 8.756C21.351 6.489 19.528 4.501 17.448 4.501C14.92 4.501 13.508 7.437 13.496 7.466C13.266 8.028 12.34 8.028 12.109 7.466C12.095 7.436 10.684 4.501 8.15501 4.501H8.154Z"
            fill="#536471"
          />
        </svg>
        <span
          class="sc-cSiKfk gTDNyI"
        >
          8
        </span>
      </div>
    </div>
  </div>
</div>
`;
